### PR TITLE
Add `PBXProject` prefix `PBXProj` partial generator

### PR DESCRIPTION
**Pull request by brentleyjones**

### CODE REVIEW

**Comment by brentleyjones** [comment]
@MobileNativeFoundation/rules_xcodeproj-maintainers I might not merge this for a bit, but I want us to have discussions about it now.

This is the first of the `.pbxproj` partial generators. They will be used in a new incremental project generation. It is the smallest, least complicated, etc. _But_ it will be used as a blueprint to copy for all of the other generators. So I want to use this PR as a place to discuss _everything_ about how the generator is coded, from its use of `ArgumentParser`, to how the tests are set up, to how things are documented, etc. Nothing is off limits, and now is the time to nit to your hearts content.

File: tools/pbxproj_generators/BUILD
Comment (brentleyjones) [inline tools/pbxproj_generators/BUILD:8]: I was using this pattern still, because I thought I would add fixtures for these... but I don't think we will be using fixtures by the time all of this is done. So I'll probably move those values into this file.

File: tools/pbxproj_generators/lib/PBXProj/src/Identifiers.swift
Comment (brentleyjones) [inline tools/pbxproj_generators/lib/PBXProj/src/Identifiers.swift:1]: I need to add a doc comment to this, because it's not clear why this is needed at first glance.

File: tools/pbxproj_generators/lib/PBXProj/src/Logger.swift
Comment (brentleyjones) [inline tools/pbxproj_generators/lib/PBXProj/src/Logger.swift:1]: Copied (with formatting changes) from `generator`. I'll back-port the formatting changes.

File: tools/pbxproj_generators/lib/PBXProj/src/SemanticVersion.swift
Comment (brentleyjones) [inline tools/pbxproj_generators/lib/PBXProj/src/SemanticVersion.swift:1]: Copied (with formatting changes) from `generator`. I'll back-port the formatting changes.

File: tools/pbxproj_generators/project_prefix/src/Generator/Generator.swift
Comment (brentleyjones) [inline tools/pbxproj_generators/project_prefix/src/Generator/Generator.swift:17]: Does this need a doc comment?

File: tools/pbxproj_generators/project_prefix/test/CalculateProjectPrefixTests.swift
Comment (brentleyjones) [inline tools/pbxproj_generators/project_prefix/test/CalculateProjectPrefixTests.swift:20]: This uses tabs for the indent, which is expected in the output. You can't easily see that though, and I bet it can be a pain for development. Currently I opted for here and in `Generator.calculateProjectPrefix()` to directly use them, since `\t` is hard to read, and developing an indent system seemed like a lot of overhead when we will always know the exact indent we want.

I'm open to feedback though. Maybe we just need a comment pointing out that it uses tabs and it's intentional?

File: tools/pbxproj_generators/project_prefix/src/Generator/Arguments.swift
Comment (brentleyjones) [inline tools/pbxproj_generators/project_prefix/src/Generator/Arguments.swift:9]: Everywhere I mention `.pbxproj` I can probably say `PBXProj` instead. A downside to that is besides for the filename it's never called that (unlike the other nested types like `PBXProject`). If people search though, they will find it called that in `XcodeProj` and the like.

```suggestion
Path to where the 'project_prefix' '.pbxproj' partial should be written.
```

File: tools/pbxproj_generators/lib/PBXProj/src/SemanticVersion.swift
Comment (CognitiveDisson) [inline tools/pbxproj_generators/lib/PBXProj/src/SemanticVersion.swift:93]: That variable name ( `self0` ) is a little confusing and doesn't make things clear to me.

Comment (CognitiveDisson) [reply tools/pbxproj_generators/lib/PBXProj/src/SemanticVersion.swift:93]: Maybe something like `instance` is a bit better

File: tools/pbxproj_generators/project_prefix/src/Generator/Generator.swift
Comment (brentleyjones) [inline tools/pbxproj_generators/project_prefix/src/Generator/Generator.swift:20]: Since we can't have labels for closures, I feel this adds to readability.

File: tools/pbxproj_generators/project_prefix/src/ProjectPrefix.swift
Comment (brentleyjones) [reply tools/pbxproj_generators/project_prefix/src/ProjectPrefix.swift]: Now that these are under the `//tools/generators` package, the generators probably need a more descriptive name.

Comment (brentleyjones) [reply tools/pbxproj_generators/project_prefix/src/ProjectPrefix.swift]: Added a README that expands on the above. Still probably need better names.

File: tools/generators/pbxproject_prefix/README.md
Comment (brentleyjones) [inline tools/generators/pbxproject_prefix/README.md:1]: I'll fill this file in soon (Monday at this point). It will detail the inputs and show an example output.

File: tools/generators/README.md
Comment (thiagohmcruz) [inline tools/generators/README.md:39]: Nit: It was a bit confusing to see "all X except for Y" for something called "prefix", I'd expect a "prefix" to generate a piece (the start) of the thing and not almost "all" of it. This is explicitly stated in `pbxproject_prefix/README.md` and IMO a similar wording could be used here

File: tools/generators/pbxproject_prefix/README.md
Comment (thiagohmcruz) [inline tools/generators/pbxproject_prefix/README.md:24]: I've never used `ArgumentParser`, is it common there to use positional args like this and not specify something like `--output-path` instead? Or that is an option and we're choosing to not make that the default usage?

File: tools/generators/pbxproject_prefix/src/Generator/CalculatePBXProjectPrefix.swift
Comment (thiagohmcruz) [inline tools/generators/pbxproject_prefix/src/Generator/CalculatePBXProjectPrefix.swift:5]: IIUC this is really "generating" the prefix partial and not "calculating" it correct?

File: tools/generators/pbxproject_prefix/src/Generator/CalculateProjectDir.swift
Comment (thiagohmcruz) [inline tools/generators/pbxproject_prefix/src/Generator/CalculateProjectDir.swift:8]: See other comment on using the word "calculate", why not simply call it `projectDir`? No strong preference here btw, feel free to ignore!

File: tools/generators/pbxproject_prefix/src/Generator/Generator.swift
Comment (thiagohmcruz) [inline tools/generators/pbxproject_prefix/src/Generator/Generator.swift:10]: >It can be used to generate multiple partials.

Multiple partials of the same `pbxproject_prefix` "type" I assume correct? Given the `generate()` impl it looks like this is specific to `pbxproject_prefix`. 

Are we planning to implement a `Generator` class like this for all partial generators?

File: tools/generators/pbxproject_prefix/test/CalculateCompatibilityVersionTests.swift
Comment (thiagohmcruz) [inline tools/generators/pbxproject_prefix/test/CalculateCompatibilityVersionTests.swift:50]: What if a user pass `14.2.0.14C18`? Is this something we should handle/error out nicely if it's not expected?

**Review by thiagohmcruz — APPROVED** [review]
Overall LGTM, I have some questions but all non-blocking, mostly for my own learning about intent vs impl approach.

File: tools/generators/README.md
Comment (thiagohmcruz) [reply tools/generators/README.md:4]: Just saw this: https://github.com/MobileNativeFoundation/rules_xcodeproj/pull/2155#issuecomment-1536354474

Yeah, tricky one to name I guess

Comment (brentleyjones) [reply tools/generators/README.md:4]: When the incremental generator is finished, `tools/generator` will no longer exist. I can instead move it into `tools/generators/legacy` or something?

File: tools/pbxproj_generators/project_prefix/test/CalculateProjectPrefixTests.swift
Comment (thiagohmcruz) [reply tools/pbxproj_generators/project_prefix/test/CalculateProjectPrefixTests.swift]: I wouldn't over engineer it as well, IMO a comment with details should be enough

File: tools/generators/README.md
Comment (brentleyjones) [reply tools/generators/README.md:39]: What different wording? They seem identical to me.

File: tools/generators/pbxproject_prefix/README.md
Comment (brentleyjones) [reply tools/generators/pbxproject_prefix/README.md:24]: I'm only using flags like `--output-path` if required, to reduce Starlark memory usage. Times when they are required:

- Optional values
- To indicate the end of one list of values and the start of another

There is a slight readability loss, but the only place you should be constructing these are in the custom scheme definition (where there are comments), or in the Starlark code that creates the `args()` object (which will also have comments).

File: tools/generators/pbxproject_prefix/src/Generator/CalculatePBXProjectPrefix.swift
Comment (brentleyjones) [reply tools/generators/pbxproject_prefix/src/Generator/CalculatePBXProjectPrefix.swift:5]: In my mind generation is the whole thing, which is the combination of calculating it and then writing it. It's not generated until it's written to disk.

File: tools/generators/pbxproject_prefix/src/Generator/CalculateProjectDir.swift
Comment (brentleyjones) [reply tools/generators/pbxproject_prefix/src/Generator/CalculateProjectDir.swift:8]: I'm open to better names. Is `projectDir(executionRoot:)` clear? I'm bad at function names, but if this is correct Swift style, I'm all for it!

File: tools/generators/pbxproject_prefix/src/Generator/Generator.swift
Comment (brentleyjones) [reply tools/generators/pbxproject_prefix/src/Generator/Generator.swift:10]: In practice you wouldn't use the same instance of `PBXProjectPrefix.Generator` to generate multiple partials, but it's stateless, so you could.

> Are we planning to implement a `Generator` class like this for all partial generators?

Yes. Each generator tool will have its own `Generator` struct.

File: tools/generators/pbxproject_prefix/test/CalculateCompatibilityVersionTests.swift
Comment (brentleyjones) [reply tools/generators/pbxproject_prefix/test/CalculateCompatibilityVersionTests.swift:50]: That will never be sent to this tool. We force a specific format for this version here: https://github.com/MobileNativeFoundation/rules_xcodeproj/blob/925d111ccfb8d65f7701e352ae6daeddf9e61b96/xcodeproj/internal/xcodeproj_rule.bzl#L122-L129

File: tools/generators/pbxproject_prefix/src/Generator/CalculateProjectDir.swift
Comment (BalestraPatrick) [reply tools/generators/pbxproject_prefix/src/Generator/CalculateProjectDir.swift:8]: Since this function doesn't have any side effects, I think simply calling it `projectDir` is fine.

**Review by BalestraPatrick — APPROVED** [review]

File: tools/generators/README.md
Comment (brentleyjones) [reply tools/generators/README.md:4]: I moved it.

Comment (thiagohmcruz) [reply tools/generators/README.md:4]: Ok, moving to `legacy` works to address what I mentioned above, thx!

File: tools/generators/README.md
Comment (thiagohmcruz) [reply tools/generators/README.md:39]: My confusion was because I don't see the word "all" in`pbxproject_prefix/README.md`:
> The `pbxproject_prefix` generator creates a `PBXProj` partial containing the
start (i.e. prefix) of the `PBXProject` element.

But I'm overcomplicating this, this `README` here does mention `start` in the first item and then details what to expect specifically: all `PBXProject` except for X and Y

I'm going to close this nvm, not even a nit tbh haha

File: tools/generators/pbxproject_prefix/README.md
Comment (thiagohmcruz) [reply tools/generators/pbxproject_prefix/README.md:24]: >There is a slight readability loss, but the only place you should be constructing these are in the custom scheme definition (where there are comments), or in the Starlark code that creates the args() object (which will also have comments).

Ok, perfect, this addresses exactly what I was concerned about and agree that it should be fine then

File: tools/generators/pbxproject_prefix/src/Generator/CalculateProjectDir.swift
Comment (thiagohmcruz) [reply tools/generators/pbxproject_prefix/src/Generator/CalculateProjectDir.swift:8]: +1

File: tools/generators/pbxproject_prefix/src/Generator/Generator.swift
Comment (thiagohmcruz) [reply tools/generators/pbxproject_prefix/src/Generator/Generator.swift:10]: Any value in making these `Generator` structs adopt a protocol so we know "a partial Generator" is expected to be able to "init with env, generate, <maybe more expected actions in the future?>" 

Probably not something for this PR, though, just wondering if it could end up evolving this way and requiring a common interface

File: tools/generators/pbxproject_prefix/test/CalculateCompatibilityVersionTests.swift
Comment (thiagohmcruz) [reply tools/generators/pbxproject_prefix/test/CalculateCompatibilityVersionTests.swift:50]: 👍

File: tools/generators/pbxproject_prefix/src/Generator/Generator.swift
Comment (brentleyjones) [reply tools/generators/pbxproject_prefix/src/Generator/Generator.swift:10]: I personally don't see a value in that. There is only a separation between `PBXProjectPrefix` and `Generator`, because we can't set `environment` in `PBXProjectPrefix.run()`. The slip makes it so `PBXProjectPrefix` parses the arguments, then those arguments get passed to `Generator`. The environment and arguments will be tied to each generator. Each command (`PBXProjectPrefix.run` in this case) will work with exactly one generator, so no need for an interface. Also, the command and generator will be the only ones in their respective module/tool.

So yes, there will be multiple commands and generators, but that's just because there will be multiple tools. Each tool could technically do whatever it wants though, it's just that we will be following the pattern set here as a common architecture.

File: tools/generators/pbxproject_prefix/src/Generator/CalculatePBXProjectPrefix.swift
Comment (brentleyjones) [reply tools/generators/pbxproject_prefix/src/Generator/CalculatePBXProjectPrefix.swift]: Went with `calculate()` for this and `write()` for the other function. Since they are the primary functions of the generator, this kinda makes it clearer (by removing a confusing second part of the name).

File: tools/generators/pbxproject_prefix/src/Generator/ProjectDir.swift
Comment (mattrobmattrob) [inline tools/generators/pbxproject_prefix/src/Generator/ProjectDir.swift:11]: Mind adding a comment about why you're removing `/private` in this case?

File: tools/generators/pbxproject_prefix/src/Generator/Calculate.swift
Comment (mattrobmattrob) [reply tools/generators/pbxproject_prefix/src/Generator/Calculate.swift:16]: Seemingly, it's just so it always has a new line at the bottom?

**Review by mattrobmattrob — APPROVED** [review]
IMO, this is going to drastically improve the understandability of the repo. 👍

File: tools/generators/README.md
Comment (brentleyjones) [reply tools/generators/README.md:68]: Yes.

File: tools/generators/lib/PBXProj/src/Identifiers.swift
Comment (brentleyjones) [reply tools/generators/lib/PBXProj/src/Identifiers.swift:6]: I'll be sure to do that as I add more.

File: tools/generators/lib/PBXProj/src/String+Extensions.swift
Comment (brentleyjones) [reply tools/generators/lib/PBXProj/src/String+Extensions.swift:19]: Lifted from `XcodeProj`, so I'm not sure tbh.

File: tools/generators/pbxproject_prefix/BUILD
Comment (brentleyjones) [reply tools/generators/pbxproject_prefix/BUILD:68]: Not anymore

File: tools/generators/pbxproject_prefix/src/Generator/Arguments.swift
Comment (brentleyjones) [reply tools/generators/pbxproject_prefix/src/Generator/Arguments.swift:21]: It's because the path is generated during the execution phase, so we can't pass the value itself as an argument. This is because we can't get absolute paths in the analysis phase.

File: tools/generators/pbxproject_prefix/src/Generator/Calculate.swift
Comment (brentleyjones) [reply tools/generators/pbxproject_prefix/src/Generator/Calculate.swift:16]: Needs the newline at the bottom, because of how the final `cat` will work. All the partials need trailing newlines (like real unix text files).

File: tools/generators/README.md
Comment (luispadron) [inline tools/generators/README.md:9]:
```suggestion
file. This file details the project itself, and controls almost all of Xcode’s
```

File: tools/generators/README.md
Comment (luispadron) [inline tools/generators/README.md:1]: Really helpful!

**Review by luispadron — APPROVED** [review]
Sweet!

File: tools/generators/pbxproject_prefix/src/Generator/Calculate.swift
Comment (brentleyjones) [reply tools/generators/pbxproject_prefix/src/Generator/Calculate.swift:16]: I'll add that as a comment.

File: test/fixtures/generator/BUILD
Comment (brentleyjones) [inline test/fixtures/generator/BUILD:2]: I'm going to land the legacy rename as its own PR.

**Pull request merged**
